### ISSUE TITLE
[FEATURE] Detect if changes need to be applied by "terramate generate"

### ISSUE THREAD

**Issue by cwe1ss** [terramate-io/terramate#1312]
**Is your feature request related to a problem? Please describe.**
I want to ensure that developers run `terramate generate` before committing to git by using pre-commit-hooks. I can already run `terraform generate` in a hook, but I've not seen an option to fail the cmd if anything has been changed. Developers might therefore not notice the updated files and commit the outdated versions.

**Describe the solution you'd like**
Similar to the already existing `terramate fmt --check`, we could have `terramate generate --check` which does not actually change any files, but exits with:
* 0 = generated files are up to date (no changes)
* 1 = files need to be changed

The disadvantage with this option is, that people have to manually call `terramate fmt` afterwards to fix the problem.

Another option would be to have `terramate generate -detailed-exitcode` (like with [terraform plan](https://developer.hashicorp.com/terraform/cli/commands/plan#detailed-exitcode)) which does change files and returns:
* 0 = generated files are up to date (no changes)
* 1 = Error
* 2 = Succeeded with changes

This way, the hook would already run `terramate generate` (so developers don't have to do it manually) but the hook would still fail. This way, developers only have to check the updated files.

**Describe alternatives you've considered**
I could capture the output of `terramate generate` and parse it to figure out if anything has been changed and fail in that case. Seems error-prone because I'd have to rely on text output

**Comment by cwe1ss** [comment]
I played around with this a little bit yesterday and my current script for the `generate` pre-commit-hook looks like this:

```sh
OUTPUT=$(terramate generate)
if [ "$OUTPUT" = "Nothing to do, generated code is up to date" ]; then
  exit 0
else
  echo "$OUTPUT"
  exit 1
fi
```
This will always run the generate command, but if there were changes, it will fail the hook and output the changes. But this is obviously not great since it relies on the text output to stay the same.

For the `fmt`-hook I'm using the following script:
```sh
OUTPUT=$(terramate fmt --check)
RESULT=$?

if [ $RESULT -eq 1 ]; then
  terramate fmt
fi

exit $RESULT
```
Using "--check" allows me to check the result code, which is nicer than comparing the output text. If there are changes to be made, I'm failing the hook but I'm also still running "terramate fmt" so that developers don't have to invoke it by themselves.

With this logic, I get similar behavior to some built-in hooks of pre-commit like [end-of-file-fixer](https://github.com/pre-commit/pre-commit-hooks?tab=readme-ov-file#end-of-file-fixer) that also change files and fail the hook.

But both of my scripts don't look great IMO. Having the described `--detailed-exitcode` flag in `terramate fmt` & `terramate generate` would reduce both scripts to a one-liner.

**Comment by mariux** [comment]
The `terramate fmt --detailed-exit-code` and `terramate generate --detailed-exit-code` options have been merged to `main` and will be available in the next release.

**Comment by cwe1ss** [comment]
Closed since this is implemented (thank you very much!!!), but see #1649 for an open bug.
